fix cloud event flaky unit tests by adding waitgroup to fakeclient

Yongxuanzhang opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

This commit adds waitgroup to fakeclient to avoid that some goroutines are not done when we want to collect the events.
The tests are flaky because the cloud events are sent with goroutine
but we don't wait until all goroutines done to check the events. So it
is possible that some events are not collected. The waitGroup will count when each goroutine is created and decrease the count when the goroutine is done.
This change has no impact on current code.

/kind flake

Signed-off-by: Yongxuanzhang yongxuanzhang@google.com

related issues:
https://github.com/tektoncd/pipeline/issues/5160
to reproduce this issue, adding time.Sleep(100 * time.Millisecond) to https://github.com/tektoncd/pipeline/blob/b648a5b2eaebf0ece7199dcf2bd89c65137973d5/pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go#L53  
related PR:
https://github.com/tektoncd/pipeline/pull/5313
Recent failing tests:

- https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/5666/pull-tekton-pipeline-unit-tests/1582795530906374144
- https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/5662/pull-tekton-pipeline-unit-tests/1583441668340715520
- https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/5647/pull-tekton-pipeline-unit-tests/1583130831700889600

<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Fix cloud event flacky unit tests by adding EventSender
```

Yongxuanzhang commented:
cc @lbernick @vdemeester @imjasonh

lbernick commented:
based on [this comment](https://github.com/tektoncd/pipeline/issues/4404#issuecomment-985656290) from @vdemeester  it seems like we're thinking of moving cloudevents functionality out of pipelines, so it might not make sense to combine these two packages into one, but I'm not sure if this comment is still true. @afrittoli can you comment on this?

lbernick commented on pkg/reconciler/events/cloudevent/cloud_event_controller.go:
```diff
@@ -144,7 +145,9 @@ func SendCloudEventWithRetries(ctx context.Context, object runtime.Object) error
 	_, isRun := object.(*v1alpha1.Run)
 
 	wasIn := make(chan error)
+	e.WaitGroup.Add(1)
```
One thing that I'm noticing here is that the interface is really just a wrapper around a waitGroup, i.e. it is pretty similar to the original implementation that just used a plain waitGroup. I'm sorry if I led you down a rabbit hole with my suggestions about how this interface should be implemented, but now that I'm seeing it in a PR we might just want a wrapper around the CloudEventClient. Something like

```
type OurOwnCloudEventClient struct {
   client *cloudEvents.Client
   eventCount int
}

func initialize(ce *cloudEvents.Client) *OurOwnCloudEventClient {
   client = ce
   eventCount = 0
}

func (c *OurOwnCloudEventClient) SendCloudEventWithRetries(ctx, object) {
  <do some stuff>
  eventCount += 1
  go func() {
    c.client.Send()
  }
}

func (c *OurOwnCloudEventClient) checkEventsOrdered(t *testing.T, <some other args>) {
    eventsSent := []string{}
    for i := 0; i < c.eventCount, i ++ {
      e <- c.client.Events
      eventsSent = append(eventsSent, e)
    }
   // Compare expected list of events with actual list of events
}
```
In this example, eventCount can also be replaced with a waitgroup. WDYT?

Yongxuanzhang commented on pkg/reconciler/events/eventsender/events_order_check.go:
```diff
@@ -24,10 +24,11 @@ import (
 	t.Helper()
 	// Sleep 50ms to make sure events have delivered
 	time.Sleep(50 * time.Millisecond)
+	e.Wait()
```
I'm not sure if the fix can 100% replace the sleep here (Though I believe it is). Maybe if we could wait the fix is merged and then remove the sleep in another PR?

Yongxuanzhang commented:
/retest

Yongxuanzhang commented on pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:
```diff
@@ -72,7 +79,70 @@ func (c FakeClient) StartReceiver(ctx context.Context, fn interface{}) error {
+
+// CheckCloudEventsUnordered checks that all events in wantEvents, and no others,
+// were received via the given chan, in any order.
+func (c *FakeClient) CheckCloudEventsUnordered(t *testing.T, eventChan chan string, testName string, wantEvents []string) error {
```
oh yes!!

Yongxuanzhang commented on pkg/reconciler/events/cloudevent/cloud_event_controller.go:
```diff
@@ -144,7 +144,11 @@ func SendCloudEventWithRetries(ctx context.Context, object runtime.Object) error
 
 	wasIn := make(chan error)
+
+	// fix flacky cloud client related units
```
ok I will remove this

Yongxuanzhang commented on pkg/reconciler/events/cloudevent/cloudevent.go:
```diff
@@ -79,7 +79,11 @@ func (t TektonEventType) String() string {
-type CEClient cloudevents.Client
+type CEClient interface {
+	cloudevents.Client
+	AddCount()
+	DecreaseCount()
```
yes, thanks!

Yongxuanzhang commented on pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:
```diff
@@ -72,7 +79,70 @@ func (c FakeClient) StartReceiver(ctx context.Context, fn interface{}) error {
+// expects to receive. The events can be received in any order. Any extra or too few
+// events are both considered errors.
+func eventsFromChannelUnordered(c chan string, wantEvents []string) error {
+	timer := time.NewTimer(10 * time.Millisecond)
```
I think we can not remove it, it is not related to the blocking, but a timer used to read from channel

lbernick replied:
Why do the tests hang if the timer is not used? The only reason this should happen is if the expected number of events is less than the actual number of events received.

Yongxuanzhang replied:
It's because the previous code will loop +1 times, and during the extra loop since we don't have any events from channel, and also we don't have a timer, it will just hand there,
```
loop len(expected) + 1 times to catch extra erroneous events received that the test is not expecting
```

Yongxuanzhang replied:
To fix it we can just reduce the loop to len(expected) not len(expected)+1. But it would be a change to this function's behaviour. I don't know why or who implement this logic. I suggest we if we want to do it, we should leave it to another clean up PR to avoid extra changes in this PR

lbernick replied:
Oh I see-- the reason is because we don't want to make sure extra events have been sent, so we wait a little extra to see if any more events are received. What we really need is for the fake client to be able to indicate that it's done sending events (e.g. by closing the channel), but the fake client doesn't know when events are done being sent because they can only be sent one at a time so I'm not sure if this is possible. Let me think about if there's a better way to do this. Keeping this logic here means that our tests are still not deterministic.

Yongxuanzhang replied:
Sure! Thanks!

lbernick replied:
ok! I have an idea! We should update [newFakeClient](https://github.com/tektoncd/pipeline/blob/14eba83bf31277dc3f8c91fa798b38e30d0a8e6a/pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go#L43-L48) to initialize the fake client buffer size with len(wantEvents). 

The reason is that sends block on a buffered channel when it is full. We are currently waiting (via waitGroup.Wait()) until all events are sent before reading any events from the channel. This means if the controller sends more events than expected, these sends will block because the channel is full, and Wait() will wait indefinitely.

With this solution, if the client sends more or fewer events than expected, Wait() will block indefinitely, but if the client sends the correct number of events, Wait() will wait until they are all sent, and no longer! a.k.a. no more flakes!

Does this make sense? WDYT? If you implement this please remove the timer and add some commentary on this strategy.

Yongxuanzhang replied:
Maybe? Let me test this solution. Thank you for coming up with this idea!!

Yongxuanzhang replied:
I tried to implement this in the latest commit, plz lmk if that is aligned what you're suggesting.

Yongxuanzhang commented on pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:
```diff
@@ -72,7 +79,70 @@ func (c FakeClient) StartReceiver(ctx context.Context, fn interface{}) error {
+// events are both considered errors.
+func eventsFromChannelUnordered(c chan string, wantEvents []string) error {
+	timer := time.NewTimer(10 * time.Millisecond)
+	expected := append([]string{}, wantEvents...)
```
I tried this method and it probably doesn't work, in this function it will do regex matching from expected events to got events, the reason is the events from channel is much more complex so we cannot simply do diff between two list of strings.

lbernick replied:
Can you try defining your own comparison function? Here's an [example](https://github.com/tektoncd/pipeline/blob/main/pkg/internal/computeresources/compare/compare.go#L56-L79)

Yongxuanzhang replied:
It basically means we will reuse the current code to compare the two list? If so I don't think it adds any values here. And it may has less readability than the current version.

Yongxuanzhang replied:
Sure! Let me give it a try. Thanks for your advice!

Yongxuanzhang replied:
I think the possible issue is that, we cannot guarantee that the two sorted lists have the same order. 
e.g.:
[a,b,c,] -> [a,b,c]
[za,by,xc]->[by,xc,za]
If we iterate them and compare on by one then this is not working.

Yongxuanzhang commented:
/retest

Yongxuanzhang commented:
/retest

lbernick commented on pkg/reconciler/events/cloudevent/cloudevent.go:
```diff
@@ -78,8 +78,15 @@ func (t TektonEventType) String() string {
+// and has methods to count goroutines that send cloud events
+type CEClient interface {
+	cloudevents.Client
+	// addCount can be implemented to add count for goroutine
+	addCount()
+	// addCount can be implemented to decrease count for goroutine
+	decreaseCount()
+}
```
```suggestion
	// addCount increments the count of events to be sent
	addCount()
	// decreaseCount decrements the count of events to be sent, indicating the event has been sent
	decreaseCount()
}
```

Yongxuanzhang replied:
Thanks!

lbernick commented on pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:
```diff
@@ -72,7 +79,69 @@ func (c FakeClient) StartReceiver(ctx context.Context, fn interface{}) error {
+	c.waitGroup.Add(1)
+}
+
+// DecreaseCount decreases 1 count for waitGroup
+func (c FakeClient) decreaseCount() {
+	c.waitGroup.Done()
+}
+
```
I wonder if `decreaseCount` is even necessary? you could call c.waitGroup.Done() within`Send`

Yongxuanzhang replied:
Probably not? Because the fakeclient `Send` is not only used in `SendCloudEventWithRetries`'s gorountine. There is another function `SendCloudEvents` also used it and it doesn't need the waitgroup to block.

Yongxuanzhang replied:
I'm not sure since that's confusing. 😂 We don't have gorontine in `SendCloudEvents `

lbernick reviewed (approved):
Thanks @Yongxuanzhang for addressing this flake that has been annoying everyone since... before I joined Tekton :) Approving since this fixes the flake, and the codebase is improved by this change, but please make sure to address the readability and style comments before merging.

Yongxuanzhang commented on pkg/reconciler/events/cloudevent/cloudevent.go:
```diff
@@ -88,8 +88,15 @@ func (t TektonEventType) String() string {
-// CEClient matches the `Client` interface from github.com/cloudevents/sdk-go/v2/cloudevents
-type CEClient cloudevents.Client
+// CEClient wraps the `Client` interface from github.com/cloudevents/sdk-go/v2/cloudevents
+// and has methods to count goroutines that send cloud events
```
Sure!

Yongxuanzhang commented on pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:
```diff
@@ -37,13 +40,17 @@ type FakeClient struct {
 
 // newFakeClient is a FakeClient factory, it returns a client for the target
-func newFakeClient(behaviour *FakeClientBehaviour) cloudevents.Client {
+func newFakeClient(behaviour *FakeClientBehaviour, clientChanbuffersize int) CEClient {
```
yesm thanks!

Yongxuanzhang commented on pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:
```diff
@@ -72,7 +79,66 @@ func (c FakeClient) StartReceiver(ctx context.Context, fn interface{}) error {
 // WithClient adds to the context a fake client with the desired behaviour
-func WithClient(ctx context.Context, behaviour *FakeClientBehaviour) context.Context {
-	return context.WithValue(ctx, ceKey{}, newFakeClient(behaviour))
+func WithClient(ctx context.Context, behaviour *FakeClientBehaviour, clientChanbuffersize int) context.Context {
```
sure!

Yongxuanzhang commented on pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:
```diff
@@ -72,7 +79,66 @@ func (c FakeClient) StartReceiver(ctx context.Context, fn interface{}) error {
+	// fakeclient's channel buffersize equals to the size of wantEvents, so no extra events will be sent
+	maxEvents := len(expected)
+	for eventCount := 0; eventCount < maxEvents; eventCount++ {
+		select {
```
removed! Thanks!

afrittoli commented on pkg/reconciler/events/cloudevent/cloudeventclient.go:
```diff
@@ -59,7 +62,38 @@ func withCloudEventClient(ctx context.Context) context.Context {
+	return context.WithValue(ctx, ceKey{}, celient)
+}
+
+// CloudClient is wrapper of cloud client and implements AddCount and DecreaseCount
```
NIT: CloudClient is a wrapper of CloudEvents client and implements addCount and decreaseCount

Yongxuanzhang replied:
Thank you!

afrittoli commented on pkg/reconciler/events/cloudevent/cloudevent.go:
```diff
@@ -88,8 +88,15 @@ func (t TektonEventType) String() string {
 
-// CEClient matches the `Client` interface from github.com/cloudevents/sdk-go/v2/cloudevents
-type CEClient cloudevents.Client
+// CEClient wraps the `Client` interface from github.com/cloudevents/sdk-go/v2/cloudevents
+// and has methods to count the cloud events being sent
```
NIT: perhaps it's worth commenting that the extra methods are for testing purposes only

Yongxuanzhang replied:
Sure!

afrittoli commented on pkg/reconciler/customrun/customrun_test.go:
```diff
@@ -187,7 +188,7 @@ func TestReconcile_CloudEvents(t *testing.T) {
 			ceClient := clients.CloudEvents.(cloudevent.FakeClient)
-			err = eventstest.CheckEventsUnordered(t, ceClient.Events, tc.name, tc.wantCloudEvents)
+			ceClient.CheckCloudEventsUnordered(t, tc.name, tc.wantCloudEvents)
 			if err != nil {
 				t.Errorf(err.Error())
 			}
```
There is no `err` returned anymore to check!

Yongxuanzhang replied:
thanks! I have fixed all of them

afrittoli commented on pkg/reconciler/customrun/customrun_test.go:
```diff
@@ -196,7 +197,7 @@ func TestReconcile_CloudEvents(t *testing.T) {
 			}
-			err = eventstest.CheckEventsUnordered(t, ceClient.Events, tc.name, []string{})
+			ceClient.CheckCloudEventsUnordered(t, tc.name, []string{})
 			if err != nil {
 				t.Errorf(err.Error())
 			}
```
There is no `err` returned anymore to check!

afrittoli commented on pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:
```diff
@@ -37,13 +38,17 @@ type FakeClient struct {
 	return FakeClient{
 		behaviour: behaviour,
-		Events:    make(chan string, bufferSize),
+		// set buffersize to length of want events to make sure no extra events are sent
```
What happens if the controller tried to send more events than expected? Would the test fail?

Yongxuanzhang replied:
Yes it will fail and this is also the previous behaviour.

Yongxuanzhang replied:
Oh I see the problem, this will lead to deadlock, I added a buffer size check before sending the event, if extra event wants to be sent it will return error. Plz lmk if that works, thanks!

afrittoli replied:
OK - it sounds good - it's important for the test to fail in that case otherwise we would miss error conditions.
Since the test method there is not trivial, I wonder if we should have a test that verifies this case.

Yongxuanzhang replied:
Sure! I will add a test for this!

Yongxuanzhang replied:
Added a test for this, ptal! 🙏

lbernick commented:
/hold cancel
since this is ready for review

lbernick commented:
@Yongxuanzhang could you please update your commit message title + body to match the PR description?

Yongxuanzhang commented:
> @Yongxuanzhang could you please update your commit message title + body to match the PR description?

Oh thanks! I just updated it. Sorry I made some changes after @afrittoli's reviews. you may want to remove your approval and review again? 

1. I added a check in `Send` so if the channel is full we will return error, this should prevent the extra events.
2. I loop the channel size not the len(expectedEvents), this will avoid blocking on reading more events when we send less events. This should work in this PR since we will wait to ensure no events are missed

lbernick commented on pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:
```diff
@@ -72,7 +83,67 @@ func (c FakeClient) StartReceiver(ctx context.Context, fn interface{}) error {
+// eventsFromChannelUnordered takes a chan of string and a list of events that a test
+// expects to receive. The events can be received in any order. Any extra or too few
+// events are both considered errors.
+func eventsFromChannelUnordered(c chan string, wantEvents []string) error {
```
It might make sense to merge this function with `CheckCloudEventsUnordered`, since it's not used anywhere else as far as I can tell, and it needs the call to waitgroup.Wait in order to work correctly.

Yongxuanzhang replied:
👍

lbernick commented on pkg/reconciler/events/cloudevent/cloudeventsfakeclient_test.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+func TestSend_Error(t *testing.T) {
+	sendEvents := []event.Event{{
```
nit: this doesn't need to be a slice

Yongxuanzhang replied:
👍

lbernick commented on pkg/reconciler/events/cloudevent/cloudeventsfakeclient.go:
```diff
@@ -72,7 +83,67 @@ func (c FakeClient) StartReceiver(ctx context.Context, fn interface{}) error {
+func eventsFromChannelUnordered(c chan string, wantEvents []string) error {
+	expected := append([]string{}, wantEvents...)
+	channelEvents := len(c)
+	// fakeclient's channel buffersize equals to the size of wantEvents, so no extra events will be sent
```
This comment is slightly misleading. Buffered channels can send once events have been pulled off of them. The reason no extra events will be sent is because the fake client returns an error in this case, but that logic lives far away from here. 

I'd suggest adding commentary to the fake client where you return an error (to explain why this prevents extra events from being sent) and some commentary here about why this block detects the case where too few events are sent.

Yongxuanzhang replied:
👍

afrittoli commented on pkg/reconciler/events/cloudevent/cloudeventsfakeclient_test.go:
```diff
@@ -0,0 +1,74 @@
+	fakeClient := Get(ctx).(FakeClient)
+
+	// the channel size is 0 so no more events can be sent
+	err := fakeClient.Send(ctx, sendEvent)
+	if err == nil {
+		t.Fatalf("want err but got nil")
+	}
```
Thank you 🙏

afrittoli commented on pkg/reconciler/testing/logger.go:
```diff
@@ -24,11 +24,15 @@ import (
+}
+
+// SetupFakeCloudClientContext sets up the fakeclient to context
+func SetupFakeCloudClientContext(ctx context.Context, expectedEventCount int) context.Context {
```
NIT: Why does this need to be a separate function?

afrittoli commented on test/events/events.go:
```diff
@@ -35,19 +35,6 @@ func CheckEventsOrdered(t *testing.T, eventChan chan string, testName string, wa
 	return nil
```
NIT: It's a bit inconsistent now that the unordered check has been moved inside to the fake client but the ordered one is in this test module. It might we worth keeping this logic in the same place, for maintenance and to help test developers. BTW, do you think the ordered check suffers from the same flake as the other one?

afrittoli replied:
This is something we could fix as a follow-up.

Yongxuanzhang replied:
Yes I'm looking into the k8s event as well, but I don't think they are the same root cause. 
Some prs:
https://github.com/tektoncd/pipeline/pull/5752
https://github.com/tektoncd/pipeline/pull/5770

afrittoli reviewed (commented):
Thank you!
/lgtm
